File: 프로그래머스/0/120878. 유한소수 판별하기/유한소수 판별하기.py
def solution(a, b):
    # 기본적으로 무한소수라고 가정하고 2로 설정
    answer = 2

    # 유클리드 호제법을 사용하여 a와 b의 최대공약수(GCD) 구하기
    x = a
    y = b

    while y:
        x, y = y, x % y  # GCD 계산

    # 분모를 GCD로 나누어 기약분수의 분모 구하기
    b //= x

    # 분모에서 2를 제거
    while b % 2 == 0:
        b //= 2

    # 분모에서 5를 제거
    while b % 5 == 0:
        b //= 5

    # 남은 값이 1이면 유한소수, 그렇지 않으면 무한소수
    if b == 1:
        answer = 1

    return answer  # 최종 결과 반환

# ✅ 예제 테스트 실행
print(solution(7, 20))
# 결과: 1 (7/20 = 0.35 → 유한소수)

print(solution(11, 22))
# 결과: 1 (11/22 = 0.5 → 유한소수)

print(solution(12, 21))
# 결과: 2 (12/21 = 0.571428... → 무한소수)

File: 프로그래머스/0/120878. 유한소수 판별하기/유한소수 판별하기.java
public class Solution {

    public static int solution(int a, int b) {
        int answer = 2;

        // 최대공약수(GCD)를 구함
        int gcd = getGCD(a, b);

        // 기약분수의 분모를 계산
        b /= gcd;

        // 분모에서 2 제거
        while (b % 2 == 0) {
            b /= 2;
        }

        // 분모에서 5 제거
        while (b % 5 == 0) {
            b /= 5;
        }

        // 남은 값이 1이면 유한소수
        if (b == 1) {
            answer = 1;
        }

        return answer;
    }

    // 유클리드 호제법으로 GCD 계산
    private static int getGCD(int x, int y) {
        while (y != 0) {
            int temp = x % y;
            x = y;
            y = temp;
        }
        return x;
    }

    // ✅ 테스트용 main
    public static void main(String[] args) {
        System.out.println(solution(7, 20));   // 출력: 1
        System.out.println(solution(11, 22));  // 출력: 1
        System.out.println(solution(12, 21));  // 출력: 2
    }
}
